spec/app/domain/authentication/authenticate_spec.rb

```diff
@@ -0,0 +1,172 @@
+
+  def authenticator(pass:)
+    double('Authenticator').tap do |x|
+      allow(x).to receive(:valid?).and_return(pass)
+    end
+  end
+
+  def input(
+    authenticator_name: 'authn-always-pass',
+    service_id: nil,
+    account: 'my-acct',
+    username: 'my-user',
+    password: 'my-pw',
+    origin: '127.0.0.1',
+    request: nil
+  )
+    Authentication::Input.new(
+      authenticator_name: authenticator_name,
+      service_id: service_id,
+      account: account,
+      username: username,
+      password: password,
+      origin: origin,
+      request: request
+    )
+  end
+
+  let (:authenticators) do
+    {
+      'authn-always-pass' => authenticator(pass: true),
+      'authn-always-fail' => authenticator(pass: false)
+    }
+  end
+
+  ####################################
+  # Security doubles
+  ####################################
+
+  let (:mocked_security_validator) {double("MockSecurityValidator")}
+  let (:mocked_origin_validator) {double("MockOriginValidator")}
+
+  before(:each) do
+    allow(Authentication::ValidateSecurity)
+      .to receive(:new)
+            .and_return(mocked_security_validator)
+    allow(mocked_security_validator).to receive(:call)
+                                          .and_return(true)
+
+    allow(Authentication::ValidateOrigin)
+      .to receive(:new)
+            .and_return(mocked_origin_validator)
+    allow(mocked_origin_validator).to receive(:call)
+                                        .and_return(true)
+  end
+
+  ####################################
+  # ENV doubles
+  ####################################
+
+  let (:two_authenticator_env) do
+    {'CONJUR_AUTHENTICATORS' => 'authn-always-pass, authn-always-fail'}
+  end
+
+  let (:blank_env) {Hash.new}
+
+  ####################################
+  # TokenFactory double
+  ####################################
+
+  # NOTE: For _this_ class, the details of actual Conjur tokens are irrelevant
+  #
+  let (:a_new_token) {'A NICE NEW TOKEN'}
+
+  let (:token_factory) do
+    double('TokenFactory', signed_token: a_new_token)
+  end
+
+  #  ____  _   _  ____    ____  ____  ___  ____  ___
+  # (_  _)( )_( )( ___)  (_  _)( ___)/ __)(_  _)/ __)
+  #   )(   ) _ (  )__)     )(   )__) \__ \  )(  \__ \
+  #  (__) (_) (_)(____)   (__) (____)(___/ (__) (___/
+
+
+  context "An unavailable authenticator" do
+    subject do
+      input_ = input(
+        authenticator_name: 'AUTHN-MISSING'
+      )
+
+      Authentication::Authenticate.new.(
+        authenticator_input: input_,
+          authenticators: authenticators,
+          enabled_authenticators: two_authenticator_env,
+          token_factory: token_factory
+      )
+    end
+
+    it "raises AuthenticatorNotFound" do
+      expect {subject}.to raise_error(
+                            Authentication::AuthenticatorNotFound
+                          )
+    end
+  end
+
+  context "An available authenticator" do
+    context "that receives invalid credentials" do
+      subject do
```

Comment:
I think we should remove the `input` helper function, and instead explicitly create the input we want in each test.  We want the tests to be fully independent (here they are coupled via the helper function `input`) and to be locally readable (here we have to scroll back up to `input` to remind ourselves what the default values are when reading.  This is a case where duplication is actually desirable.

Reply:
got it